# Contreras-Lab
Code I've written for my research position in the Contreras Lab at UT Austin as a high school student.

### Code use:
Each folder contains a different project.
Some folders contain subfolders. These are for when the same base code had to be run on multiple different datasets and required some modification.

Please see .py files inside folders for code and descriptions about what the code does.
The rest of the files in each folder are inputs or outputs.

### Project summaries:
<b>Application of fluorescence-based in vivo RNA regional accessibility data to identify novel sRNA targets</b>\
sRNA target misidentification is attributed to heavy reliance of computational predictions on free energy predictions and lack of in vivo insight. Informing computational predictions with in vivo accessibility data can help elucidate potential targets.
See "" for code.

<b>Integrative data-mining approach that couples DNA protein occupancy, RNA-seq, and RNA accessibility data with motif identification and target prediction algorithmss</b>\

##### Application of fluorescence-based in vivo RNA regional accessibility data to identify novel sRNA targets

### Publications:
Mihailovic, M. K., Ekdahl, A., Chen, A., Leistra, A. N., Li, B., Javier González Martínez, Law, M., Ejindu, C., Massé, E., Freddolino, P. L., & Contreras, L. M. (2021). Uncovering Transcriptional Regulators and Targets of sRNAs Using an Integrative Data-Mining Approach: H-NS-Regulated RseX as a Case Study. <i>Frontiers in Cellular and Infection Microbiology, 11</i>. https://doi.org/10.3389/fcimb.2021.696533

Bowman, E. R., Mihailovic, M. K., Li, B., & Contreras, L. M. (2020). Bioinformatic Application of Fluorescence-Based In Vivo RNA Regional Accessibility Data to Identify Novel sRNA Targets. In V. Arluison & F. Wien (Eds.), <i>RNA Spectroscopy</i> (pp. 41–71). Humana. https://doi.org/10.1007/978-1-0716-0278-2_5

